File: src/com/leetcode/backtracking/M31_Next_Permutation.java
package com.leetcode.backtracking;

import java.util.Arrays;

/**
 * Created by Michael on 2017/1/23.
 *
 * Implement next permutation, which rearranges numbers into the lexicographically next greater permutation of numbers.
 * If such arrangement is not possible, it must rearrange it as the lowest possible order (ie, sorted in ascending order).
 * Note: The replacement must be in-place, do not allocate extra memory.
 *
 * Here are some examples. Inputs are in the left-hand column and its corresponding outputs are in the right-hand column.
 * Since the total permutations for {1, 2, 3} are:
 *  ↓ 1 2 3 ←
 *  ↓ 1 3 2  |
 *  ↓ 2 1 3  |
 *  ↓ 2 3 1  |
 *  ↓ 3 1 2  |
 *  ↓ 3 2 1 →
 *
 * Thus,
 * 1,2,3 next → 1,3,2
 * 3,2,1 next → 1,2,3
 * 1,1,5 next → 1,5,1
 *
 * Function Signature:
 * public void nextPermutation(int[] a) {...}
 *
 * <Tags>
 * - 利用已排序数组的性质，即使是部分已排序的数组也很有用。
 *
 */
public class M31_Next_Permutation {
    public static void main(String[] args) {
        int[] a = {5, 1, 1};
        for (int i = 0; i < 24; i++) {
            nextPermutation(a);
            System.out.println(Arrays.toString(a));
        }
    }

    /** 解法2：找规律 (以右为上) + 局部交换 + 局部反转。Time - o(n), Space - o(1). */
    // 首先要搞清楚Lexicographic顺序排列的特征是<以右为上>。即尽可能从最右端开始交换元素。
    // 例如{1,2,3}的下一个排列是{1,3,2}，而不是{2,1,3}
    // 例如{3,7,2,5}的下一个排列是{3,7,5,2}，而不是{7,3,2,5}
    // 然后针对三种情况进行分析：
    // Case#1 如果数组是<完全降序>排列的（相等元素也算降序），例如{7,4,2}，说明当前状态已经是最后一个Permutation，需要回到第一个Permutation
    // Case#2 如果数组是<完全升序>排列的，例如{3,4,6,9}，那么只需要交换最后两个元素即可得到下一个Permutation。
    // Case#3 如果数组是<部分降序>排列的，需要分成三步运算：
    /** 第一步：定位。找到从右向左<首次违反降序排列>的元素位置i。*/
    // 例：{6 8 | 9 7 4 1} a[1] = 8 是第一个违反降序的元素
    //    {1 4 2 8 5 | 7} a[4] = 5 是第一个违反降序的元素
    //    {8 | 9 8 6 4 2} a[0] = 8 是第一个违反降序的元素
    /** 第二步：交换。从最右边向左找到<第一个比i大的元素j>，将i与j的值交换。*/
    // 这么做本质是要找到i右侧数组中比i大的最小元素，由于i右侧都是降序排列的，因此只要从右向左扫描，遇到的第一个比i大的元素就一定是比i大的最小元素。
    // 交换元素后，i右侧区间内的元素一定依然是完全降序排列的，因为把i与第一个比i大的元素交换，本质上就是把i插入了右侧降序数组中并确保整个数组依然降序排列的操作。
    // 例：{6 8 | 9 7 4 1} -> {6 9 | 8 7 4 1}
    //    {1 4 2 8 5 | 7} -> {1 4 2 8 7 | 5}
    //    {8 | 9 8 6 4 2} -> {9 | 8 8 6 4 2}
    /** 第三步：反转。将i右侧区间的<所有元素按轴对称反转>。*/
    // 这么做其实是为了让i的右侧区间恢复升序排列，以得到下一个Permutation。由于上面已经知道交换元素之后右侧区间是完全降序排列的，因此并不需要真正的排序，只需要翻转即可。
    // 例：{6 9 | 8 7 4 1} -> {6 9 | 1 4 7 8}
    //    {1 4 2 8 7 | 5} -> {1 4 2 8 7 | 5}
    //    {9 | 8 8 6 4 2} -> {9 | 2 4 6 8 8}
    static void nextPermutation(int[] a) {
        if (a == null || a.length < 2) return;
        int i = a.length - 2;                       // 让i从倒数第二个元素开始向前扫描，这样只有数组是完全降序排列时，i才会停在-1
        int j = a.length - 1;
        while (i >= 0 && a[i] >= a[i + 1]) i--;     // 第一步：定位i。只有当a[i]小于a[i+1]时i才会停下来。
        if (i >= 0) {
            while (a[j] <= a[i]) j--;               // 第二步：定位j。
            swap(a, i, j);                          // 第二步：交换i和j。
        }
        reverse(a, i + 1, a.length - 1);        // 第三步：翻转i的右侧区间。
    }

    private static void reverse(int[] a, int i, int j) {
        while (i < j) swap(a, i++, j--);
    }

    private static void swap(int[] a, int i, int j) {
        int temp = a[i];
        a[i] = a[j];
        a[j] = temp;
    }

    /** 解法1：Value-As-Index + Backtracking + 匹配。 Time - o(n!). */
    // 最简单的思路是根据数组的元素生成全排列，然后将数组于全排列每一个比较，找到自己在全排列中的位置，返回下一个排列即可，
    // 但是这样的空间复杂度和实践复杂度都很高。因为全排列的个数于数组长度之间是阶乘的关系，比指数上升还要快。
}

File: src/com/leetcode/backtracking/M31_Next_Permutation.cc
#include "../common.h"
#include <algorithm>
#include <climits>
using namespace std;

/**
 * Created by Michael on 2019/8/26.
 *  
 * Implement next permutation, 
 * which rearranges numbers into the lexicographically next greater permutation of numbers.
 * If such arrangement is not possible, it must rearrange it as the lowest possible order (ie, sorted in ascending order).
 * The replacement must be in-place and use only constant extra memory.
 * Here are some examples. 
 * Inputs are in the left-hand column and its corresponding outputs are in the right-hand column.
 * 
 * 1,2,3 → 1,3,2
 * 3,2,1 → 1,2,3
 * 1,1,5 → 1,5,1
 * 
 * 基本问题
 *      - 拿到题目，第一个问题就是，如何定义下一个排列的顺序，所谓的 Lexicographically 实际是按什么排列的？
 *      - 输入数组是否有序？无序
 *      - 输入数组元素是否无重复？可以有重复
 *      - Permutation 虽然解是循环出现的，但是最后一个状态向第一个状态的变化不再是渐变，而是突变。这个需要特殊处理。
 *      - 从函数的接口能够观察到其实是原位运算，也就暗示了我们其实要做 swap 和 reverse 来实现
 *      - 如何高效的分析问题：举例。这个题目的主要难点在于如果不自己列一遍3位或者4位的所有全排列，你完全想象不到从一个状态到下一个状态可能出现哪些奇葩的变化特点，所以必须要多举例，而且善于从例子中提炼规律，这个过程确实很考验人。
 * 
 * 基本思想
 *      - 大智若愚：这道题属于真的没有什么捷径的题，需要踏踏实实的观察相邻 permutation 的特性，找规律，用代码实现出来。
 *      - 虽然这道题被归为 backtracking，但是实际上用 backtracking 来解这个的复杂度是 O(n!)，而完全靠分析相邻排列的变化规律，不使用任何特殊 technique 就可以做出来，复杂度只有 O(n)
 *      - element swap 
 *      - element reverse
 *      - reverse scan of array
 *      - exploit sorted array: 虽然整个数组无序，但是可以利用有序的部分，快速确定相对大小元素的索引位置。比如升序的部分，最大的元素一定位于最后。
 * 
 */

void mySwap(vector<int> a, int i, int j) {
    a[i] = a[i] ^ a[j];
    a[j] = a[i] ^ a[j];
    a[i] = a[i] ^ a[j];
}

void myReverse(vector<int>& nums) {
    for (int i = 0; i < nums.size() / 2; i++) {
        mySwap(nums, i, nums.size() - i - 1);
    }
}

// 第一步：反向扫描数组，找到第一个不满足降序排列的元素，记为 pivot。
// 第二步：如果数组严格降序，就反转整个数组，返回结果，其复杂度是 O(n)（并不需要排序，因为排序是 O(nlogn)）
// 第三步：在 pivot 之后的元素中，选择比 pivot 大的最小元素，与 pivot 交换。（注意：这里用到了部分已排序数组的性质！利用规律！）
void nextPermutation(vector<int>& nums) {
    if (nums.size() < 2) return;            // 空数组和单个元素的数组永远只有一个排列方式，就是自己。
    int i = nums.size() - 2;                // 确保 i 能够停在第一个不满足降序条件的元素上（而不是它旁边的）
    while (i >= 0) {
        if (nums[i] < nums[i + 1]) break;   // 找到第一个不满足降序排列的数组就立即退出，i 停留在这个元素的位置上不再动了
        i--;
    }
    if (i < 0) {                            // i < 0 说明始终没有 break，说明数组是严格降序排列的，也就是整个排列中的最后一个状态，此时需要反转整个数组，获得严格升序的数组就回到第一状态了。
        reverse(nums.begin(), nums.end());
        return;
    }
    // 到此位置，一定存在 pivot
    int pivot = nums[i];
    for (int j = nums.size() - 1; j > i; j--) {     // 逆序扫描，由于后面部分的元素一定是降序排列，因此最小的元素一定出现在最后。
        if (nums[j] > pivot) {                      // 只要元素大于 pivot 就是满足最小且大于 pivot 的值，直接 swap 即可结束。
            swap(nums[i], nums[j]);
            break;
        }
    }
    reverse(nums.begin() + i + 1, nums.end());      // 反转 pivot 之后区间的元素  [2, 1, 4, 3] ---swap---> [2, 3, 4, 1] ---reverse---> [2, 3, 1, 4]
}

/**
 *      示例分析
 * 
 *             大于 pivot 且最小的元素                               反转 pivot 之后的区域
 *             ↓
 *    1, 4, 3, 2               ------------>        2,  |  4, 3, 1    ---------->    2, 1, 3, 4
 *    ↑
 *    第一个不满足降序的元素 pivot
 * 
 */


int main() {
    vector<int> a = {2, 3, 1, 4};
    for (int i = 0; i < 16; i++) {
        nextPermutation(a); print(a);
    }
}